src/common/mempool.cc

```diff
@@ -12,21 +12,156 @@
+// Statically assign threads to shards to minimize how many threads use the
+// same shard
+static int thread_shard_next;
+static std::mutex thread_shard_mtx;
+
+#else
+// Use sched_getcpu() to determine which core a thread is running on,
+// use this to assign the thread a shard and cache this value to
+// eliminate overhead of repeatedly calling sched_getcpu. Periodically
+// check the cached value to deal with non-bound threads switching cores.
+// Ideally this means that each CPU core accesses its own shard and there
+// are no shared CPU cachelines. If a thread switches cores performance
+// will be non-optimal (but atomics ensure correctness) until the switch
+// is noticed and a new shard is selected
+
+enum {
+  thread_shard_recheck_min_interval = 1,
+  thread_shard_recheck_max_interval = 10000
+};
+
+static thread_local int thread_shard_recheck_count;
+static thread_local int thread_shard_recheck_interval;
+
+// Keep statistics on how successfully we manage to track which core a thread
+// is running on
+//
+// correct_core - number of times thread to core binding is correctly tracked
+// incorrect_core - number of times thread to core binding is incorrectly tracked
+// checked_core - number of times sched_getcpu() has been called
+//
+// sched_getcpu_cost = 32 cycles
+// cacheline_ping_pong_cost = 150 cycles (intra socket), 400 cycles (inter socket)
+//
+// These costs will vary between CPU architectures, but relative difference between
+// costs is likely to be similar
+//
+// cached_algorithm_cost = checked_core * sched_getcpu_cost + incorrect_core * cacheline_ping_pong_cost
+// uncached_algorithm_cost = (correct_core + incorrect_core) * sched_getcpu_cost
+
+struct fault_stats_t {
+  ceph::atomic<size_t> correct_core = {0};
+  ceph::atomic<size_t> incorrect_core = {0};
+  ceph::atomic<size_t> checked_core = {0};
+
+  char __padding[128 - sizeof(ceph::atomic<size_t>)*2];
+
+  void dump(ceph::Formatter *f) const {
+    f->dump_int("correct_core", correct_core );
+    f->dump_int("incorrect_core", incorrect_core);
+    f->dump_int("checked_core", checked_core);
+  }
+
+  fault_stats_t& operator+=(const fault_stats_t& o) {
+    correct_core += o.correct_core;
+    incorrect_core += o.incorrect_core;
+    checked_core += o.checked_core;
+    return *this;
+  }
+} __attribute__ ((aligned (128)));
+
+static fault_stats_t *shard_faults;
+#endif
+
+}
+
+size_t mempool::get_num_shards(void) {
+  if (num_shards==0) {
+    size_t threads = std::thread::hardware_concurrency();
+    if (threads == 0) {
+      threads = default_shards;
+    }
+    if (threads < min_shards) {
+      threads = min_shards;
+    }
+    if (threads > max_shards) {
+      threads = max_shards;
+    }
+    threads--;
+    while (threads!=0) {
+      num_shard_bits++;
+      threads>>=1;
+    }
+    num_shards = 1 << num_shard_bits;
+#if defined(MEMPOOL_SCHED_GETCPU)
+    shard_faults = new fault_stats_t[num_shards];
```

Comment:
Wonder if we need a destructor for this allocation, (or a std:shared_ptr smart pointer to deallocate itself, like in https://stackoverflow.com/questions/14993791/in-c-how-to-write-a-destructor-for-freeing-memory-of-pointer-to-a-structure ).  Also, are we allocating each time that the function get_num_shards() is called? Shall we allocate only once instead (e.g. we might need to initialise the static pointer to say, nullptr, and only allocate a new array if it has not been allocated already)?

Reply:
This is one off initialization, so will only happen the 1st time get_num_shards() is called - num_shards starts off as 0 and gets set to a non-zero value inside this if the first time the function is called.

Agree that shard_faults should probably be a std::shared_ptr